From c304aaec3201284a0122472499409b84b558a8b1 Mon Sep 17 00:00:00 2001
From: Subhendu Sekhar Behera <sbehera@broadcom.com>
Date: Fri, 11 Apr 2014 19:47:04 +0530
Subject: [PATCH 1444/1532] nae: Fix timer based PHY autonegotiation.

Remove the use of nae_id & keep port_index counter independent.
Also fix the autonegotiation case; i.e. do autonegotiation when
cable is plugged-in; not the other way.
[Based on SDK 3.2]
Signed-off-by: Quanyang Wang <quanyang.wang@windriver.com>
---
 drivers/net/ethernet/broadcom/nae/xlpge_nae.c | 20 +++++++++++++++-----
 1 file changed, 15 insertions(+), 5 deletions(-)

diff --git a/drivers/net/ethernet/broadcom/nae/xlpge_nae.c b/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
index 428fbf2..79e1205 100644
--- a/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
+++ b/drivers/net/ethernet/broadcom/nae/xlpge_nae.c
@@ -803,23 +803,33 @@ void nlm_xlp_nae_remove(void)
 
 static int nlm_xlp_port_poll(void *buff)
 {
-	uint32_t i,num_ports,maxnae,node,speed,duplex,port_index;
+	uint32_t i,num_ports,maxnae,node,speed,duplex;
 	nae_t *nae_cfg;
 	static int arr[NLM_MAX_NODES][18];
 	maxnae = get_num_nae_pernode();
+	for (node = 0;node < NLM_MAX_NODES; node++) {
+		int num_nae, tmp, port_index=0;
+		for(num_nae=0; num_nae<maxnae; num_nae++) {
+			nae_cfg = get_nae(node, num_nae);
+			if (nae_cfg == NULL)
+				continue;
+			for(i=0; i < nae_cfg->num_ports; i++, port_index++) {
+				arr[node][port_index]=netsoc_get_phy_status(&nae_cfg->ports[i],&speed,&duplex);
+			}
+		}
+	}
 	while(1) {
 		set_current_state(TASK_UNINTERRUPTIBLE);
 		for (node = 0;node < NLM_MAX_NODES; node++) {
-			int num_nae, tmp;
+			int num_nae, tmp, port_index=0;
 			for(num_nae=0; num_nae<maxnae; num_nae++) {
 				nae_cfg = get_nae(node, num_nae);
 				if (nae_cfg == NULL)
 					continue;
-				for(i=0; i < nae_cfg->num_ports; i++) {
-					port_index = i + (nae_cfg->nae_id * nae_cfg->num_ports);
+				for(i=0; i < nae_cfg->num_ports; i++, port_index++) {
 					tmp=arr[node][port_index];
 					arr[node][port_index]=netsoc_get_phy_status(&nae_cfg->ports[i],&speed,&duplex);
-					if ((arr[node][port_index]>-1)&&(arr[node][port_index]<tmp))
+					if ((arr[node][port_index]>-1)&&(arr[node][port_index] > tmp))
 						netsoc_start_autoneg(&nae_cfg->ports[i]);
 					else
 						continue;
-- 
1.9.1

